synapse/storage/databases/main/user_directory.py

```diff
@@ -105,8 +105,10 @@ def _make_staging_area(txn: LoggingTransaction) -> None:
                 GROUP BY room_id
             """
             txn.execute(sql)
-            rooms = [{"room_id": x[0], "events": x[1]} for x in txn.fetchall()]
-            self.db_pool.simple_insert_many_txn(txn, TEMP_TABLE + "_rooms", rooms)
+            rooms = list(txn.fetchall())
+            self.db_pool.simple_insert_many_txn(
+                txn, TEMP_TABLE + "_rooms", keys=("room_id", "events"), values=rooms
+            )
             del rooms
```

Comment:
I wondered if I could just set `values=txn.fetchall()`, but I recalled there being something sketchy with iterating a `txn` as you start another query? Figured it was safer to cast it to a list first.

Reply:
I thought `list(txn)` and `txn.fetchall()` were equivalent to be honest. Doubt it matters but this looks like it's building the list twice needlessly. See https://www.python.org/dev/peps/pep-0249/#fetchall (Twisted notes that it's compatible with the standard DBAPI).